include/cifer/abe/dippe.h

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright (c) 2018 XLAB d.o.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CIFER_DIPPE_H
+#define CIFER_DIPPE_H
+
+#include <amcl/ecp2_BN254.h>
+#include <amcl/fp12_BN254.h>
+#include <gmp.h>
+#include <stdlib.h>
+
+#include "cifer/data/mat.h"
+#include "cifer/data/mat_curve.h"
+#include "cifer/data/vec.h"
+#include "cifer/data/vec_curve.h"
+#include "cifer/internal/errors.h"
+
+/**
+* \file
+* \ingroup abe
+* \brief This is a decentralized inner-product predicate encryption
+* scheme by Yan Michalevsky and Marc Joye: "Decentralized Policy-Hiding
+* Attribute-Based Encryption with Receiver Privacy"
+*/
+
+/**
+ * cfe_dippe represents the DIPPE scheme
+ */
+typedef struct cfe_dippe {
+    size_t assump_size;
+    mpz_t p;
+    cfe_mat_G1 g1_A;
+    cfe_mat_G1 g1_UA;
+} cfe_dippe;
+
+/**
+ * Initializes the DIPPE scheme and generates the global parameters
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param assump_size The size of the underlying assumption
+ */
+void cfe_dippe_init(cfe_dippe *dippe, size_t assump_size);
+
+/**
+ * Clears the scheme and frees allocated memory
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ */
+void cfe_dippe_free(cfe_dippe *dippe);
+
+/**
+ * cfe_dippe_attribute_vector represents an attribute vector that is used
+ * to request attributes from an authority
+ */
+typedef struct cfe_dippe_attribute_vector {
+    char *s;
+    size_t len;
+    mpz_t attrs;
+} cfe_dippe_attribute_vector;
+
+/**
+ * Prepares the attribute vector struct
+ *
+ * @param av A pointer to a cfe_dippe_attribute_vector struct; Represents the resulting attribute vector
+ * @param vec_len Length of the resulting attribute vector
+ * @param pattern Array of indices to define the attribute vector
+ * @param pat_len Length of the pattern array
+ * @return Error code
+ */
+cfe_error cfe_dippe_attribute_vector_init(cfe_dippe_attribute_vector *av, size_t vec_len, size_t pattern[], size_t pat_len);
+
+/**
+ * Clears the attribute vector and frees allocated memory
+ *
+ * @param av A pointer to a cfe_dippe_attribute_vector struct
+ */
+void cfe_dippe_attribute_vector_free(cfe_dippe_attribute_vector *av);
+
+/**
+ * cfe_dippe_policy_vector represents a policy vector that is used
+ * during encryption
+ */
+typedef struct cfe_dippe_policy_vector {
+    size_t len;
+    cfe_vec pol;
+} cfe_dippe_policy_vector;
```

Comment:
`cfe_vec` already has an atribute `len`, so as I see this struct is not needed.

Reply:
You're right. I've removed this struct.